### PR TITLE
DP-860 - Email address validation made GDS-compliant

### PR DESCRIPTION
**Pull request by maciej-goaco**

### CODE REVIEW

File: Libraries/CO.CDP.Mvc.Validation/EmailAddressValidator.cs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+public static class EmailAddressValidator
+{
+    private static readonly string ValidLocalChars = @"a-zA-Z0-9.!#$%&'*+/=?^_`{|}~\-";
+    private static readonly string EmailRegexPattern = $@"^[{ValidLocalChars}]+@([^.@][^@\s]+)$";
+    private static readonly Regex HostnamePartRegex = new(@"^(xn|[a-z0-9]+)(-?-[a-z0-9]+)*$", RegexOptions.IgnoreCase);
+    private static readonly Regex TldPartRegex = new(@"^([a-z]{2,63}|xn--([a-z0-9]+-)*[a-z0-9]+)$", RegexOptions.IgnoreCase);
+
+    public static bool IsValid(string emailAddress)
+    {
+        emailAddress = StripAndRemoveObscureWhitespace(emailAddress);
```
Comment (jakzal) [inline Libraries/CO.CDP.Mvc.Validation/EmailAddressValidator.cs:17]: If I understand this correctly, we strip some characters for validation. Does it mean we then persist the email with those characters included?

**Review by andymantell — APPROVED** [review]

File: Libraries/CO.CDP.Mvc.Validation/EmailAddressValidator.cs (hunk 1 above)
Comment (maciej-goaco) [reply Libraries/CO.CDP.Mvc.Validation/EmailAddressValidator.cs]: Sanitising the value before validation has been taken out of the validation process and applied separately with a ModelBinder. This way it has become reusable and when modified to be applicable on the Page model level, we could easily apply it solution-wide to improve security (as fallback for eliminating XSS, SQL injections etc. if any sneaked in due to not using MS HTML extensions).

File: Frontend/CO.CDP.OrganisationApp.Tests/Extensions/StringExtensionsTests.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+using FluentAssertions;
+
+namespace CO.CDP.OrganisationApp.Tests.Extensions;
+
+public class StringExtensionsTests
+{
+    [Theory]
+    [InlineData("validated-text")]
+    [InlineData("  \tvalidated-text \x0c ")]
+    [InlineData("\rv\u200Cal\u200Did\u200Bate\u2060d-\u180Etex\uFEFFt\uFEFF")]
```
Comment (jakzal) [inline Frontend/CO.CDP.OrganisationApp.Tests/Extensions/StringExtensionsTests.cs:11]: Can we add a case for repeated characters to make sure it doesn't just remove one?

i.e. `[InlineData("ab\u200Ccd\u200Cef\u200C")]`

File: Frontend/CO.CDP.OrganisationApp/Models/SanitisedStringModelBinder.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
+using CO.CDP.OrganisationApp.Extensions;
+using Microsoft.AspNetCore.Mvc.ModelBinding;
+
+namespace CO.CDP.OrganisationApp.Models;
+
+public class SanitisedStringModelBinder : IModelBinder
```
Comment (jakzal) [inline Frontend/CO.CDP.OrganisationApp/Models/SanitisedStringModelBinder.cs:6]: 👌

**Review by jakzal — APPROVED** [review]

**Review by SDGoaco — APPROVED** [review]
Working as expected in the Local Environment

**Pull request merged**